components/datetime/src/format/datetime.rs

```diff
@@ -286,38 +326,75 @@ where
                     }
                 }
             };
-            format_number(w, value, field.length)?
+            format_number(
+                w,
+                fixed_decimal_format,
+                FixedDecimal::from(value),
+                field.length,
+            )?
         }
         FieldSymbol::Minute => format_number(
             w,
-            usize::from(
+            fixed_decimal_format,
+            FixedDecimal::from(usize::from(
                 datetime
                     .datetime()
                     .minute()
                     .ok_or(Error::MissingInputField)?,
-            ) as isize,
+            ) as isize),
             field.length,
         )?,
-        FieldSymbol::Second(Second::Second) => format_number(
-            w,
-            usize::from(
+        FieldSymbol::Second(Second::Second) => {
+            let seconds = usize::from(
                 datetime
                     .datetime()
                     .second()
                     .ok_or(Error::MissingInputField)?,
-            ) as isize,
-            field.length,
-        )?,
-        FieldSymbol::Second(Second::FractionalSecond) => format_number(
-            w,
-            usize::from(
-                datetime
-                    .datetime()
-                    .nanosecond()
-                    .ok_or(Error::MissingInputField)?,
-            ) as isize,
-            field.length,
-        )?,
+            );
+            if let Some(PatternItem::Field(next_field)) = next_item {
+                if let FieldSymbol::Second(Second::FractionalSecond) = next_field.symbol {
+                    let mut fraction = FixedDecimal::from(usize::from(
+                        datetime
+                            .datetime()
+                            .nanosecond()
+                            .ok_or(Error::MissingInputField)?,
+                    ));
+
+                    // We only support fixed field length for fractional seconds.
+                    let precision = match next_field.length {
+                        FieldLength::Fixed(p) => p,
+                        _ => {
+                            return Err(Error::FixedDecimal);
+                        }
+                    };
+
+                    fraction
+                        .multiply_pow10(-9)
+                        .map_err(|_| Error::FixedDecimal)?;
+
+                    let mut buffer = seconds.to_string();
+                    // Strip the leading zero, so we append something like .123 to the seconds
+                    buffer.push_str(fraction.to_string().get(1..).unwrap_or(""));
+
+                    let mut num =
+                        FixedDecimal::from_str(&buffer).map_err(|_| Error::FixedDecimal)?;
+                    num.pad_right(precision as u16);
```

Comment:
Thought: There's probably a better way to do this. Didn't notice this on my first pass. Let me take a look.

Reply:
I'm putting up a PR with code in FixedDecimal to make this better.

Reply:
Thanks for adding the new method to FixedDecimal :)